### ISSUE TITLE
FreeBSD: needs superuser ro run

### ISSUE THREAD

**Issue by nunotexbsd** [orf/gping#236]
gping 1.5.1 / FreeBSD

Running with a normal user, a blank window and error:
`Error: receiving on a closed channel`

Runs ok with superuser root.

**Comment by orf** [comment]
Ok, let me take a look at this.

**Comment by orf** [comment]
@nunotexbsd I've tried to make ping failures (exits etc) more graceful. Can you give master a go and tell me what happens?

Open to suggestions on how to to make this more fluid.

**Comment by Klaws--** [comment]
> Running with a normal user, a blank window and error: `Error: receiving on a closed channel`
> 
> Runs ok with superuser root.

That's the expected result. Sending ICMP packets required superuser privileges on Unix systems. The `ping` command has the setuid flag set, so it's always run on superuser level.

There are several option how gping could get around this issue.
1. Use the `ping` command and parse the output.
2. Do a UDP Ping instead of an ICMP Echo Request.
3. Ask the user to `chown `and `setuid `a precompiled gping binary.

Option 2 could be done transparently - if the script detects it can't use ICMP Echo Requests, it could try to fall back to UDP Pings.

**Comment by ehaupt** [comment]
Running `gping` `1.16.1` as unprivileged user on FreeBSD 14.0-RELEASE amd64 I get:

```sh
$ gping www.google.com
There was an error running ping: exit status: 1
Stderr: ping: Operation not permitted: only root may use interval < 1s
```

However, the suggested workaround always worked for me:

```sh
$ gping --watch-interval 1 www.google.com
[runs as expected]
```

If it is desired to have the `suid` bit set I could add an option similar to the [trippy](https://github.com/freebsd/freebsd-ports/blob/main/net/trippy/Makefile#L17-L22) port.

**Comment by orf** [comment]
I never realised how stupidly and mind-numbingly complex running the “ping” command would be before I started this project.

Sure, I can add a default interval of 1s to freebsd builds if the user is not root. Why not. Why do I have to? Reasons. Why does it have a specific and different ping output than other OSs? More reasons. Why are the flags different? You guessed it, “reasons”.

If anyone fancies doing this then the fix is fairly simple, otherwise I’ll do it the next time I have some OSS time scheduled.

**Comment by orf** [comment]
Does anyone know why the hell super user privileges are required to use an interval of less than 1s? If it’s because we’re timelocked to the 1980s and we’re trying to prevent malicious users from DDOSing machines with ping i’m going to loose my mind.

**Comment by ehaupt** [comment]
I can understand your frustration. I'm not sure as to why but I suspect the same.

Could you please review my [branch](https://github.com/ehaupt/gping/commit/f67825ed841f1556234ce9f7f94925021918b7db) I've created? It works(tm) on my machine.

**Comment by orf** [comment]
Hey @ehaupt, thanks for the branch! Sorry about the late reply, I was travelling for work and had limited time. Your branch looks great - would you be able to make a PR so you are attributed + the tests are run? If not I can cherry pick and do it myself.

**Comment by ehaupt** [comment]
Hey no worries @orf there is no rush. I'll work out a PR. I think we should probably discriminate between Linux and *BSD.

**Comment by orf** [comment]
I've also sent a message to freebsd-hackers to see what's up with the restriction 🤞 

> Hey no worries @orf there is no rush. I'll work out a PR. I think we should probably discriminate between Linux and *BSD.

I'm happy to rework the PR to handle this - we also need to ensure it works on Windows as well. I had a brief look at the `users` crate and it seems like it would need to be gated by a `cfg` macro as it uses [libc directly](https://github.com/ogham/rust-users/blob/d6883b4cc1bf00fdca920f5064ea551684fc0faf/src/lib.rs#L128). Also happy to do this if you don't have the capacity 👍

**Comment by ehaupt** [comment]
I would be more than happy if you could integrate my rather unrefined prototype. I have also linked [another PR](https://github.com/orf/gping/issues/392) (OpenBSD) to this issue. It seems that addressing this directly at the source would definitely benefit users.
